### PR TITLE
Fix logic to enable fast path on input line on Android

### PR DESCRIPTION
**Pull request by andreasavio**
To reduce input latency, the Android doc recommends to configure the audio recorder with `SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION`, which disables all inbuilt audio effects (noise cancellations, equaliser...) on the input line. JUCE doesn't do this by default and uses instead `SL_ANDROID_RECORDING_PRESET_GENERIC`. To enable the fast input line, I had to fix a couple of bugs, specifically:

- Missing `SL_IID_ANDROIDCONFIGURATION` interface in the audio recorder;
- The recorder cannot be reconfigured after being realised.

These are fixed and fast path is enabled by default. However, in the end I didn't see an improvement either in the latency or the audio detection :(

@sbergen , your call if we want to merge this.

### CODE REVIEW

**Comment by sbergen** [comment]
@aklapuri Any thoughts on this?

**Comment by aklapuri** [comment]
For me, any preset designed for speech recognition sounds like a great choice for us also, in terms of audio quality. Not to even speak of low latency. Did you test the fix on more than one device? Could it be that latency improves only in some cases or on some devices? If Android doc specifically says it is the best choice in terms of low latency, then maybe it is the best bet in the long run?
By the way, I did not see SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION in this PR, but maybe it was hidden behind the configuration variable.

**Comment by sbergen** [comment]
@aklapuri I agree. `SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION ` seems to be behind `setAudioPreprocessingEnabled`.

File: modules/juce_audio_devices/native/juce_android_OpenSL.cpp [hunk 1]
```diff
@@ -620,21 +643,26 @@ class OpenSLAudioIODevice  : public AudioIODevice,
                                                                    static_cast<SLuint32> (bufferList.numBuffers) };
             SLDataSink audioSink = { &bufferQueue, &pcmFormat };
 
-            const SLInterfaceID interfaceIDs[] = { *engine.SL_IID_ANDROIDSIMPLEBUFFERQUEUE };
+            const SLInterfaceID interfaceIDs[] = { *engine.SL_IID_ANDROIDSIMPLEBUFFERQUEUE,
+                *engine.SL_IID_ANDROIDCONFIGURATION };
```
Comment (sbergen) [inline modules/juce_audio_devices/native/juce_android_OpenSL.cpp]: @andreasavio  Would you mind to elaborate a bit on this change?

**Comment by andreasavio** [comment]
@sbergen , @aklapuri , yes, `SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION` is configured in `Recorder::setAudioPreprocessingEnabled(bool enabled)`, if `enabled==false`. This is an internal method which is called from the client-facing API `AudioIODevice::setAudioPreprocessingEnabled()`.

I did not change the internal method, because there's nothing wrong with it. There is a problem with the API method, however, as it assumes that the recorder config can be changed on the fly, while in fact the audio engine must be recycled.

File: modules/juce_audio_devices/native/juce_android_OpenSL.cpp (hunk 1 above)
Comment (andreasavio) [reply modules/juce_audio_devices/native/juce_android_OpenSL.cpp]: The `SL_IID_ANDROIDCONFIGURATION` interface provides the `SetConfiguration` method, which is used in `Recorder::setAudioPreprocessingEnabled()`. If we create the audio recorder without that interface, `Recorder::configObject` (line 657) will be `NULL` and `Recorder::setAudioPreprocessingEnabled()` will fail. It looks like a bug in JUCE... Superpowered does instantiate the audio recorder with `SL_IID_ANDROIDCONFIGURATION` enabled.

Comment (sbergen) [reply modules/juce_audio_devices/native/juce_android_OpenSL.cpp]: Ok, that's what it looked like also, just wanted to confirm. Good job finding this 👍

**Review by sbergen — APPROVED** [review]
:+1:

**Comment by andreasavio** [comment]
@aklapuri , yes, I tested this on our android lineup (4, 5, and 6). The audio flinger dump confirms that the effects are disabled on the input line.

**Pull request merged**